File: decompiled/sources/com/hjq/toast/ToastStrategy.java
package com.hjq.toast;

import android.app.Activity;
import android.app.Application;
import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.os.Message;
import com.hjq.toast.config.IToast;
import com.hjq.toast.config.IToastStrategy;
import com.hjq.toast.config.IToastStyle;
import java.lang.ref.WeakReference;

public class ToastStrategy extends Handler implements IToastStrategy {
    private static final int DELAY_TIMEOUT = 200;
    private static final int TYPE_CANCEL = 2;
    private static final int TYPE_SHOW = 1;
    private ActivityStack mActivityStack;
    private Application mApplication;
    private WeakReference<IToast> mToastReference;
    private IToastStyle<?> mToastStyle;

    public ToastStrategy() {
        super(Looper.getMainLooper());
    }

    public void registerStrategy(Application application) {
        this.mApplication = application;
        this.mActivityStack = ActivityStack.register(application);
    }

    public void bindStyle(IToastStyle<?> iToastStyle) {
        this.mToastStyle = iToastStyle;
    }

    public IToast createToast(Application application) {
        IToast iToast;
        Activity foregroundActivity = this.mActivityStack.getForegroundActivity();
        if (foregroundActivity != null) {
            iToast = new ActivityToast(foregroundActivity);
        } else if (Build.VERSION.SDK_INT == 25) {
            iToast = new SafeToast(application);
        } else {
            iToast = new SystemToast(application);
        }
        if ((iToast instanceof ActivityToast) || Build.VERSION.SDK_INT < 30 || application.getApplicationInfo().targetSdkVersion < 30) {
            iToast.setView(this.mToastStyle.createView(application));
            iToast.setGravity(this.mToastStyle.getGravity(), this.mToastStyle.getXOffset(), this.mToastStyle.getYOffset());
            iToast.setMargin(this.mToastStyle.getHorizontalMargin(), this.mToastStyle.getVerticalMargin());
        }
        return iToast;
    }

    public void showToast(CharSequence charSequence) {
        removeMessages(1);
        Message obtain = Message.obtain();
        obtain.what = 1;
        obtain.obj = charSequence;
        sendMessageDelayed(obtain, 200);
    }

    public void cancelToast() {
        removeMessages(2);
        sendEmptyMessage(2);
    }

    public void handleMessage(Message message) {
        WeakReference<IToast> weakReference = this.mToastReference;
        IToast iToast = weakReference != null ? (IToast) weakReference.get() : null;
        int i = message.what;
        if (i != 1) {
            if (i == 2 && iToast != null) {
                iToast.cancel();
            }
        } else if (message.obj instanceof CharSequence) {
            CharSequence charSequence = (CharSequence) message.obj;
            if (iToast != null) {
                iToast.cancel();
            }
            IToast createToast = createToast(this.mApplication);
            this.mToastReference = new WeakReference<>(createToast);
            createToast.setDuration(getToastDuration(charSequence));
            createToast.setText(charSequence);
            createToast.show();
        }
    }

    /* access modifiers changed from: protected */
    public int getToastDuration(CharSequence charSequence) {
        return charSequence.length() > 20 ? 1 : 0;
    }
}

File: decompiled/sources/com/hjq/toast/ToastStrategy.kt
package com.hjq.toast

import android.app.Activity
import android.app.Application
import android.os.Build
import android.os.Handler
import android.os.Looper
import android.os.Message
import com.hjq.toast.config.IToast
import com.hjq.toast.config.IToastStrategy
import com.hjq.toast.config.IToastStyle
import java.lang.ref.WeakReference

open class ToastStrategy : Handler(Looper.getMainLooper()), IToastStrategy {
    private var mActivityStack: ActivityStack? = null
    private var mApplication: Application? = null
    private var mToastReference: WeakReference<IToast>? = null
    private var mToastStyle: IToastStyle<*>? = null

    override fun registerStrategy(application: Application) {
        mApplication = application
        mActivityStack = ActivityStack.register(application)
    }

    override fun bindStyle(style: IToastStyle<*>) {
        mToastStyle = style
    }

    override fun createToast(application: Application): IToast {
        val foregroundActivity: Activity? = mActivityStack!!.getForegroundActivity()
        val toast: IToast = if (foregroundActivity != null) {
            ActivityToast(foregroundActivity)
        } else if (Build.VERSION.SDK_INT == 25) {
            SafeToast(application)
        } else {
            SystemToast(application)
        }
        if (toast is ActivityToast || Build.VERSION.SDK_INT < 30 || application.applicationInfo.targetSdkVersion < 30) {
            val style = mToastStyle!!
            toast.setView(style.createView(application))
            toast.setGravity(style.getGravity(), style.getXOffset(), style.getYOffset())
            toast.setMargin(style.getHorizontalMargin(), style.getVerticalMargin())
        }
        return toast
    }

    override fun showToast(text: CharSequence) {
        removeMessages(TYPE_SHOW)
        val message = Message.obtain().apply {
            what = TYPE_SHOW
            obj = text
        }
        sendMessageDelayed(message, DELAY_TIMEOUT.toLong())
    }

    override fun cancelToast() {
        removeMessages(TYPE_CANCEL)
        sendEmptyMessage(TYPE_CANCEL)
    }

    override fun handleMessage(message: Message) {
        val toast: IToast? = mToastReference?.get()
        when (message.what) {
            TYPE_SHOW -> {
                val text = message.obj
                if (text is CharSequence) {
                    toast?.cancel()
                    val createToast = createToast(mApplication!!)
                    mToastReference = WeakReference(createToast)
                    createToast.setDuration(getToastDuration(text))
                    createToast.setText(text)
                    createToast.show()
                }
            }

            TYPE_CANCEL -> {
                toast?.cancel()
            }
        }
    }

    protected open fun getToastDuration(text: CharSequence): Int = if (text.length > 20) 1 else 0

    private companion object {
        private const val DELAY_TIMEOUT: Int = 200
        private const val TYPE_SHOW: Int = 1
        private const val TYPE_CANCEL: Int = 2
    }
}
